daemon/unit.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * This file is part of GyroidOS
+ * Copyright(c) 2025 Fraunhofer AISEC
+ * Fraunhofer-Gesellschaft zur Förderung der angewandten Forschung e.V.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2 (GPL 2), as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GPL 2 license for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * The full GNU General Public License is included in this distribution in
+ * the file called "COPYING".
+ *
+ * Contact Information:
+ * Fraunhofer AISEC <gyroidos@aisec.fraunhofer.de>
+ */
+
+#include "unit.h"
+#include "u_user.h"
+#include "u_perm.h"
+
+#include "common/macro.h"
+#include "common/mem.h"
+#include "common/event.h"
+#include "common/fd.h"
+#include "common/list.h"
+#include "common/sock.h"
+#include "common/uuid.h"
+#include "compartment.h"
+
+#include <stdint.h>
+#include <sys/inotify.h>
+#include <sys/stat.h>
+
+struct unit {
+	compartment_t *compartment;
+	char *sock_name;
+	char *data_path;
+	event_inotify_t *inotify_sock_dir;
+	void (*on_sock_connect_cb)(int sock, const char *sock_path);
+	compartment_callback_t *state_observer;
+	bool restart;
+};
+
+static char *unit_bin_path[] = { "/usr/sbin", "/sbin", "/usr/bin", "/bin" };
+
+// list of compartment modules for all container objects
+static list_t *compartment_module_list = NULL;
+
+static list_t *
+unit_get_compartment_modules(void)
+{
+	return compartment_module_list;
+}
+
+static void
+unit_set_extension(void *extension_data, compartment_t *compartment)
+{
+	ASSERT(extension_data);
+	ASSERT(compartment);
+
+	unit_t *unit = extension_data;
+	unit->compartment = compartment;
+}
+
+const char *
+unit_get_sock_dir(const unit_t *unit)
+{
+	ASSERT(unit);
+	u_user_t *u_user = compartment_module_get_instance_by_name(unit->compartment, "c_user");
+
+	return u_user_get_sock_dir(u_user);
+}
+
+static void
+unit_do_create_cb(const char *path, uint32_t mask, UNUSED event_inotify_t *inotify, void *data)
+{
+	IF_FALSE_RETURN(mask & IN_CREATE);
+
+	unit_t *unit = data;
+	ASSERT(unit);
+
+	int sock = -1;
+
+	struct stat s;
+	mem_memset(&s, 0, sizeof(s));
+
+	if (stat(path, &s) == -1) {
+		WARN_ERRNO("Could not stat %s", path);
+		return;
+	}
+
+	IF_FALSE_RETURN((s.st_mode & S_IFMT) == S_IFSOCK);
+
+	IF_FALSE_RETURN(strstr(path, unit->sock_name));
```

Comment:
RETURN_ERROR here as S_IFSOCK but no name?

Reply:
nop, we just ignore everything what is created inside the unit's /run/socket dir except a socket with the name which belongs to the unit communication. If inside the unit some payload generates a regular file or subdir there we just don't care about. In case oth the scd there is slso some token communication generating subdirs for token control sockets which would be printed here as ERROR then.